DRILL-6486: BitVector split and transfer does not work correctly for non byte-multiple transfer lengths

bitblender opened the pull request:
Fix for the bug in BitVector splitAndTransfer. The logic for handling copy of last-n bits was incorrect for none byte-multiple transfer lengths. This PR fixes this problem and adds new tests which simulate the failing condition and also exercise other changes in this PR.

HanumathRao commented on exec/java-exec/src/test/java/org/apache/drill/exec/vector/TestSplitAndTransfer.java:
```diff
@@ -77,4 +79,84 @@ public void test() throws Exception {
+
+  int getBit(TestBitPattern pattern, int index) {
+    if (pattern == TestBitPattern.RANDOM) {
+      final int randomBit = (int) (Math.random() * 2);
```
I was just thinking if bitValue is better than randomBit?

bitblender replied:
The variable has been removed.

HanumathRao commented on exec/vector/src/main/java/org/apache/drill/exec/vector/BitVector.java:
```diff
@@ -285,35 +285,55 @@ public void transferTo(BitVector target) {
-    int offset = startIndex % 8;
-    if (offset == 0) {
+    int firstByteIndex = getByteIndex(startIndex);//byte offset of the first src byte
+    int bytesToRead = getSizeFromCount(length); //src bytes to read (including start/end bytes that might not be fully copied)
```
bytesToRead is also being used for setting target byte so is it better to use numOfBytes or something like that?

bitblender replied:
Agreed. Changing it to 'numBytesHoldingSourceBits'

HanumathRao commented:
@bitblender  Thank you for fixing the bug in this complex code. I have only few variable naming comments. LGTM.

paul-rogers commented:
Have we considered replacing BitVector with UINT1 vector? This swap was done in most places in Drill before 1.0, probably because BitVector was buggy and slow. Perhaps we should just remove it from the remaining places?

Or, if we get it to work, consider using it where it was originally intended, as the "bits vector" in Nullable types? If we do that, we have cross-version compatibility to worry about with old clients...

bitblender commented:
@paul-rogers  I have opened https://issues.apache.org/jira/browse/DRILL-6493 to follow up on your suggestion. Until that gets prioritized and completed, the changes in this PR will be necessary to make sure that BitVector splitAndTranfer() works correctly.  

ilooner commented:
+1 LGTM Thanks for adding thorough unit tests.

sohami commented on exec/vector/src/main/java/org/apache/drill/exec/vector/BitVector.java:
```diff
@@ -285,35 +285,55 @@ public void transferTo(BitVector target) {
         target.data.release();
       }
-      target.data = data.slice(firstByte, byteSize);
+      target.data = data.slice(firstByteIndex, numBytesHoldingSourceBits);
```
I think this is true if `(firstBitOffset == 0 && length%8 == 0)` `if length%8 !=0` then we are keeping reference of more bits then requested. But as memory for underlying Drillbuf is always allocated at byte level then I guess it doesn't make any difference even if we copy extra bits.

However in target vector we are setting the `valueCount` same as `length` but it's not being used while getting any value from Vector (or in `get` apis). I think we should put that check as bounds checking. The internal Drillbuf bounds checking will not catch it since length of buffer is larger than actual data which will allow accessing a value beyond `ValueCount`.

bitblender replied:
Good point. I have added a check.

sohami commented on exec/vector/src/main/java/org/apache/drill/exec/vector/BitVector.java:
```diff
@@ -285,35 +285,55 @@ public void transferTo(BitVector target) {
+        byte bitsFromLastButOneByte = 0;
+        // if start is not byte aligned then we have to copy some bits from the last full byte read in the
+        // previous loop
+        if (firstBitOffset != 0) {
```
this will always be true here.

bitblender replied:
Yes. Removed the check.

sohami commented on exec/vector/src/main/java/org/apache/drill/exec/vector/BitVector.java:
```diff
@@ -285,35 +285,55 @@ public void transferTo(BitVector target) {
+          lastButOneByte = byteIPlus1;
+          bitsFromLastButOneByte = (byte)((lastButOneByte & 0xFF) >>> firstBitOffset);
+        }
+        final int lastByte = this.data.getByte(firstByteIndex + numBytesHoldingSourceBits);
```
This can go beyond the bound of the data buffer. Consider a case where there are only 3 bytes in SrcVector which will contain bits from 0 to 23. Now in target we want to put bits from index 5 and length 17 which means last bit to be read is 21st from src vector. `firstByteIndex` will be 0 and `numBytesHoldingSourceBits` will be 3 so it will end up calling `getByte` for index 3 which is not present.

bitblender commented:
@sohami I have fixed the last-byte issue. I removed the bounds check because some of the operators are hitting the IOB, most likely because they are not setting valueCount correctly. Those issues are best fixed in a different PR, so I am removing that check.

sohami commented on exec/vector/src/main/java/org/apache/drill/exec/vector/BitVector.java:
```diff
@@ -328,7 +328,9 @@ public void splitAndTransferTo(int startIndex, int length, BitVector target) {
-        final int lastByte = this.data.getByte(firstByteIndex + numBytesHoldingSourceBits);
+
+        final int lastByte = (8 - firstBitOffset >= length % 8) ?
+                0 : this.data.getByte(firstByteIndex + numBytesHoldingSourceBits);
```
Check looks good. Can you please add the motivation of this check in the comment. Based on my understanding `length%8` gives the number of bits to be copied from last byte and `(8-firstBitOffset)` is the bits left to be copied from `lastButOneByte`. So if bits left to be copied is smaller than bits needed from last byte then we need to look at next byte in those cases.

sohami commented on exec/java-exec/src/test/java/org/apache/drill/exec/record/vector/TestValueVector.java:
```diff
@@ -560,17 +561,20 @@ public void testBitVector() {
       // Ensure unallocated space returns 0
-      assertEquals(0, accessor.get(3));
+      // Not a valid check since bounds checking is enabled
+     //assertEquals(0, accessor.get(3));
```
since you have removed the bounds check let's uncomment this assert. Also comment says unallocated space shouldn't it be unassigned index instead ?

sohami commented on exec/java-exec/src/test/java/org/apache/drill/exec/vector/TestSplitAndTransfer.java:
```diff
@@ -77,4 +79,102 @@ public void test() throws Exception {
+      final int start = startLength[0];
+      final int length = startLength[1];
+      tp.splitAndTransfer(start, length);
+      newBitVector.getMutator().setValueCount(length);
```
instead of `setValueCount` in test I think we should check if `splitAndTransfer` is setting it correctly or not ?

bitblender commented:
Yes.



On June 21, 2018 at 10:25:44 AM, Pritesh Maker (notifications@github.com<mailto:notifications@github.com>) wrote:

@bitblender<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_bitblender&d=DwMCaQ&c=cskdkSMqhcnjZxdQVpwTXg&r=HlugibuI4IVjs-VMnFvNTcaBtEaDDqE4Ya96cugWqJ8&m=hlQMTw-fCidZaB1T0_96G3hUXi6jCqarmxAWz-qfT2M&s=sTmHTnIdq5BE9dp8wuPzCyuLG4B7r8-xKX4p4f8_LX4&e=> would you be able to address the remaining review comments in time for the batch commit this week?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_apache_drill_pull_1316-23issuecomment-2D399181499&d=DwMCaQ&c=cskdkSMqhcnjZxdQVpwTXg&r=HlugibuI4IVjs-VMnFvNTcaBtEaDDqE4Ya96cugWqJ8&m=hlQMTw-fCidZaB1T0_96G3hUXi6jCqarmxAWz-qfT2M&s=w5EwdGqBkcsdB82V-FqvQSzMI1lutQexIbn2o3Wzz3I&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAaRheMUQN2UBlKgAmUXv-5FNL5bEKqCduks5t-2D9cMgaJpZM4UjZUK&d=DwMCaQ&c=cskdkSMqhcnjZxdQVpwTXg&r=HlugibuI4IVjs-VMnFvNTcaBtEaDDqE4Ya96cugWqJ8&m=hlQMTw-fCidZaB1T0_96G3hUXi6jCqarmxAWz-qfT2M&s=H9fTIsEBO8FfUfHLXSXC7a_zimEsw_LeVQIt3JHmZec&e=>.

bitblender commented:
@sohami  Addressed all your review comments. Please take a look again.
